File: java/tournament/src/main/java/Tournament.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class Tournament {
  Map<String, Team> nameToTeam = new HashMap<>();

  String printTable() {
    return "Team                           | MP |  W |  D |  L |  P\n%s"
        .formatted(
            nameToTeam.values().stream()
                .sorted(
                    Comparator.comparing(Team::getPoint)
                        .reversed()
                        .thenComparing(team -> team.name))
                .map(
                    team ->
                        "%-30s | %2d | %2d | %2d | %2d | %2d\n"
                            .formatted(
                                team.name,
                                team.winCount + team.drawCount + team.lossCount,
                                team.winCount,
                                team.drawCount,
                                team.lossCount,
                                team.getPoint()))
                .collect(Collectors.joining()));
  }

  void applyResults(String results) {
    for (String result : results.split("\n")) {
      String[] parts = result.split(";");

      String name1 = parts[0];
      nameToTeam.putIfAbsent(name1, new Team(name1));
      Team team1 = nameToTeam.get(name1);

      String name2 = parts[1];
      nameToTeam.putIfAbsent(name2, new Team(name2));
      Team team2 = nameToTeam.get(name2);

      String outcome = parts[2];
      if (outcome.equals("win")) {
        ++team1.winCount;
        ++team2.lossCount;
      } else if (outcome.equals("draw")) {
        ++team1.drawCount;
        ++team2.drawCount;
      } else if (outcome.equals("loss")) {
        ++team1.lossCount;
        ++team2.winCount;
      }
    }
  }
}

class Team {
  String name;
  int winCount;
  int drawCount;
  int lossCount;

  Team(String name) {
    this.name = name;
  }

  int getPoint() {
    return winCount * 3 + drawCount;
  }
}

File: python/tournament/tournament.py
def tally(tournament_results):
    name_to_team = {}

    for result in tournament_results:
        name1, name2, outcome = result.split(';')

        if name1 not in name_to_team:
            name_to_team[name1] = Team(name1)
        if name2 not in name_to_team:
            name_to_team[name2] = Team(name2)

        if outcome == 'win':
            name_to_team[name1].win_count += 1
            name_to_team[name2].loss_count += 1
        elif outcome == 'draw':
            name_to_team[name1].draw_count += 1
            name_to_team[name2].draw_count += 1
        elif outcome == 'loss':
            name_to_team[name1].loss_count += 1
            name_to_team[name2].win_count += 1

    return build_table(name_to_team)


def build_table(name_to_team):
    return ['Team                           | MP |  W |  D |  L |  P'] + \
        [f'{team.name:<30} | {team.matches_played:>2} | {team.win_count:>2} | {team.draw_count:>2} | {team.loss_count:>2} | {team.point:>2}'
         for team in build_sorted_teams(name_to_team)]


def build_sorted_teams(name_to_team):
    return sorted(name_to_team.values(), key=lambda team: (-team.point, team.name))


class Team:
    def __init__(self, name):
        self.name = name
        self.win_count = self.draw_count = self.loss_count = 0

    @property
    def matches_played(self):
        return self.win_count + self.draw_count + self.loss_count

    @property
    def point(self):
        return self.win_count * 3 + self.draw_count
